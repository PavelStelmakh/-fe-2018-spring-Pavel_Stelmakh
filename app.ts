import * as express from 'express';
const cors = require('cors');
import * as bodyParse from 'body-parser';
const router = require('./routes/users');

const app = express();
app.set('port', 3000);

app.use(cors());
app.options('*', cors());
app.use(bodyParse.json());
app.use(bodyParse.urlencoded({extended: false}));

app.use(express.static(__dirname));

app.use(router);

app.listen(app.get('port'), () => console.log(`listening on port ${app.get('port')}`));