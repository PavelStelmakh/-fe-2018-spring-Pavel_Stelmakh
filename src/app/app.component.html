<app-popup>Saving...</app-popup>
<div>
    <h3>
        {{'title' | translate}}
    </h3>
    <div [class.switchLang]="true">
        <button (click)="switchLang('en')">en</button>
        <button (click)="switchLang('ru')">ru</button>
    </div>
</div>
<router-outlet></router-outlet>


