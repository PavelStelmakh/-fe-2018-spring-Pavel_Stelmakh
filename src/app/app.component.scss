:host {
    width: 500px;
    margin: 30px auto;
    border-radius: 15px;
    box-shadow: 0 0 10px 3px rgba(0,0,0,0.5);
    display: block;
    background-color: rgba(0, 0, 0, 0.7);
}

div:not(.switchLang) {
    background-color: #fa5655;
    border-top-right-radius: 15px;
    border-top-left-radius: 15px;
    height: 40px;
    font-family: Verdana;
    font-size: 1.25rem;
    line-height: 40px;
    color: rgba(0, 0, 0, 0.7);
    padding-left: 40px;
}

h3, .switchLang {
    display: inline-block;
    margin: 0;    
}

button {
    margin: 0 10px;
    width: 50px;
    height: 25px;  
    color: #ccc;
}