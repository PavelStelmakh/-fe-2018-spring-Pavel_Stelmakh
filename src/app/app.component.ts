import { Component } from '@angular/core';
import { IUserForm } from './shared/IUserForm';

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.scss']
})
export class AppComponent {
  title = 'Please, fill in the fields';
  formValue: IUserForm;

  onSubmit(value: IUserForm) {
    this.formValue = value;
  }
}
