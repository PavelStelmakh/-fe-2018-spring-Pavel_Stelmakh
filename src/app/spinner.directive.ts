import { Directive, ElementRef, Renderer2 } from '@angular/core';
import { UsersService } from './users.service';

@Directive({
  selector: '[appSpinner]'
})
export class SpinnerDirective {

  constructor(
    private element: ElementRef, 
    private renderer: Renderer2,
    private userService: UsersService
    ) {
      userService.spinner.subscribe((spinner: boolean) => {
        if (spinner) {
          renderer.addClass(element.nativeElement, 'spinner');
        } else {
          renderer.removeClass(element.nativeElement, 'spinner');
        }
      });
     }

}
