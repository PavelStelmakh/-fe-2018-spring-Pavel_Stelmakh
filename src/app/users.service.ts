import { Injectable, Inject, OnDestroy } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { BASE_URL } from './url.service';
import { BehaviorSubject, Observable } from 'rxjs';
import { tap, catchError } from 'rxjs/operators';
import { User } from 'models/User';

@Injectable({
  providedIn: 'root'
})
export class UsersService implements OnDestroy {
  private id: number;
  readonly spinner: BehaviorSubject<boolean> = new BehaviorSubject<boolean>(false);

  constructor(private http: HttpClient, @Inject(BASE_URL) private url: string) {}

  setId(id: number) {
    this.id = id;
  }

  getUser() {
    this.spinner.next(true);
    return this.http.get(`${this.url}/users/${this.id}`, {
      observe: 'response',
      withCredentials: true
    })
    .pipe(
      tap(() => this.spinner.next(false))
    );
  }

  logout() {
    this.setId(0);
  }

  checkExistName(name: string, id?: number) {
    const idUser: number = id || this.id || 0;
    return this.http.get(`${this.url}/login/find/${name}/${idUser}`, {
      observe: 'response',
      withCredentials: true
    });
  }

  update(data: User, id?: number) {
    const idUser: number = id || this.id;
    return this.http.put(`${this.url}/users/${idUser}`, data, {
      observe: 'response',
      withCredentials: true
    });
  }

  addUser(data: User) {
    return this.http.post(`${this.url}/users/add`, data, {
      observe: 'response',
      withCredentials: true
    });
  }

  ngOnDestroy() {
    this.spinner.complete();
  }

  search(name: string): Observable<User[]> {
    this.spinner.next(true);
    return this.http.get<User[]>(`${this.url}/users/search/${name}`, {withCredentials: true})
    .pipe(
      tap(() => this.spinner.next(false)),
      catchError(err => {
        this.spinner.next(false);
        throw new Error;
        })
      );
  }

  deleteUser(id: number) {
    return this.http.delete(`${this.url}/users/${id}`, {withCredentials: true});
  }

}
