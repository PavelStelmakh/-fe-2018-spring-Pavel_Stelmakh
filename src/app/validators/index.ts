import { usernameValidator } from './username-validator';
import { userageValidator } from './userage-validator';
import { dateValidator } from './date-validator';
import { checkExistNameValidator } from './checkExistName-validator';

export {usernameValidator, userageValidator, dateValidator, checkExistNameValidator};