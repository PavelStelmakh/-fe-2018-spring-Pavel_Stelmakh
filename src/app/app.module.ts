import { BrowserModule } from '@angular/platform-browser';
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { Routes, RouterModule } from "@angular/router";
import { HttpClientModule, HttpClient } from '@angular/common/http';
import { TranslateModule, TranslateLoader } from '@ngx-translate/core';

import { AppComponent } from './app.component';
import { LoginModule } from './login/login.module';
import { LoginComponent } from './login/login.component';
import { BASE_URL } from './url.service';
import { AuthService } from './auth/auth.service';
import { AuthGuard } from './auth/auth.guard';
import { HttpLoaderFactory } from './shared/HttpLoaderFactory';
import { PopupModule } from './popup/popup.module';
import { SpinnerDirective } from './spinner.directive';

const route: Routes = [
  {
    path: 'user',
    loadChildren: './user/user.module#UserModule',
    canLoad: [AuthGuard]
  },
  {
    path: 'admin',
    loadChildren: './admin/admin.module#AdminModule',
    canLoad: [AuthGuard]
  },
  {
    path: 'login',
    component: LoginComponent
  },
  {
    path: '**',
    redirectTo: 'user'
  }
];

@NgModule({
  declarations: [
    AppComponent,
    SpinnerDirective  
  ],
  imports: [
    BrowserModule,    
    LoginModule,
    HttpClientModule,
    ReactiveFormsModule,
    RouterModule.forRoot(route),
    TranslateModule.forRoot({
      loader: {
        provide: TranslateLoader,
        useFactory: HttpLoaderFactory,
        deps: [HttpClient, BASE_URL]
      },
      isolate: true
    }),
    PopupModule
  ],
  providers: [
    {provide: BASE_URL, useValue: 'http://localhost:3000'},
    AuthService
  ],
  bootstrap: [AppComponent]
})
export class AppModule { }
