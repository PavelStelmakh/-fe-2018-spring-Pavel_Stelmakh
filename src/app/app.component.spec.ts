import { TestBed, async, ComponentFixture } from '@angular/core/testing';
import { RouterTestingModule } from "@angular/router/testing";
import { NO_ERRORS_SCHEMA } from '@angular/core';
import { AppComponent } from './app.component';
import { Observable, of } from 'rxjs';
import { TranslateModule, TranslateLoader, TranslateService } from '@ngx-translate/core';
import { Store } from '@ngrx/store';
import { hot } from 'jasmine-marbles';
import * as langReducer from './reducers/language.reducer';

const translations = {
  "title": "Angular training",
  "login": {
    "login": "login"
  } 
};

class FakeLoader implements TranslateLoader {
  getTranslation(lang: string): Observable<{}> {
    return of(translations);
  }
}

describe('AppComponent', () => {
  let fixture: ComponentFixture<AppComponent>;
  let translate: TranslateService;
  let store: Store<langReducer.State>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [
        RouterTestingModule,
        TranslateModule.forRoot({
          loader: {provide: TranslateLoader, useClass: FakeLoader}
        })
      ],
      declarations: [
        AppComponent
      ],
      providers: [
        {
          provide: Store,
          useValue: {
            dispatch: jest.fn(),
            pipe: jest.fn(() => hot('-a', { a: 'en' }))
          }
        }
      ],
      schemas: [
       NO_ERRORS_SCHEMA
      ]
    }).compileComponents();

  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(AppComponent);
    translate = TestBed.get(TranslateService);
    store = TestBed.get(Store);
  });

  it('renders markup to snapshot', () => {
    translate.use('en');
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

});
