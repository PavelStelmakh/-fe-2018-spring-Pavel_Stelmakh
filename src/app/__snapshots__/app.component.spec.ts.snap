// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppComponent renders markup to snapshot 1`] = `
<app-root
  store={[Function Object]}
  translate={[Function TranslateService]}
>
  <app-popup>
    Saving...
  </app-popup><div
    appspinner=""
  >
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
  </div><div
    class="content"
  >
    <h3>
       Angular training 
    </h3>
    <div
      class="LoginAndLang"
    >
      <a
        href="/login"
        ng-reflect-router-link="/login"
        routerlink="/login"
      >
        login
      </a>
      <a>
        en
      </a>
      <a>
        ru
      </a>
    </div>
  </div><router-outlet />
</app-root>
`;
